Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx

```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
   //-------------------------------------------------------
   metric->Print(std::cout);
 
+
+   //-------------------------------------------------------
+  // Check consistency between BSplineTransform
+  // derivatives computed with vs omitting weights caching.
+  //-------------------------------------------------------
+
+  // Define types/constants
+  constexpr unsigned int splineOrder = 3;
+  constexpr unsigned int nodesPerDimension = 8;
+  using BSplineTransformType = itk::BSplineTransform<CoordinateRepresentationType, ImageDimension, splineOrder>;
+  using InitializerType = itk::BSplineTransformInitializer<BSplineTransformType, MovingImageType>;
+
+  // Offset moving image to ensure derivatives for BSplineTransform are non-zero.
+  MovingImageType::PointValueType bSplineTestMovingImageOrigin[] = { 1.0f, 1.0f };
+  movingImage->SetOrigin(bSplineTestMovingImageOrigin);
+
+  // Initialise BSplineTransform
+  auto bSplineTransform = BSplineTransformType::New();
+  auto initializer = InitializerType::New();
+  BSplineTransformType::MeshSizeType meshSize;
+  meshSize.Fill(nodesPerDimension - splineOrder);
+  initializer->SetTransform(bSplineTransform);
+  initializer->SetImage(movingImage);
+  initializer->SetTransformDomainMeshSize(meshSize);
+  initializer->InitializeTransform();
+
+  // Set bSplineTransform parameters
+  ParametersType bSplineParameters(bSplineTransform->GetNumberOfParameters());
+  bSplineParameters.Fill(0.0);
```

Comment:
I am not sure whether all-zero parameters will reveal differences.

Reply:
I'm not sure I understand what you're asking here. Do you mean that checking only the consistency between cached vs non-cached derivatives is insufficient to catch differences? Or that initialising the transform with all zero parameters may return zero derivatives?

As far as the suggested test goes, the moving image offset is changed from (0,0) to (1,1) at the beginning of the test, while the fixed image keeps its offset as (0,0). If my understanding is correct, this should be sufficient to ensure derivatives are non-zero given that the interpolation weights are being used during derivative calculation. Either way, I'm happy to initialise the parameters with non-zero values?

Reply:
> initialising the transform with all zero parameters may return zero derivatives

This is what I am afraid of.

Reply:
Not sure whether I'll finish revising code changes today - I'm planning to randomly initialise each parameter with MersenneTwister. Will that be ok?